### DISCUSSION TITLE
Find papers and books about CppAD

### DISCUSSION THREAD

**Discussion by xuruilong100** [coin-or/CppAD#153] [Q&A] [answered]
I am trying to combine CppAD with [QuantLib](https://github.com/lballabio/quantlib), and it works well for some pricing engines.

Now, I have to deal with some advanced topics about AAD, for instance, implementing [special functions](https://mathworld.wolfram.com/SpecialFunction.html) and matrix factorization algorithms (LU, QR, SVD and Cholesky) by atomic functions.

I need know more details about CppAD. So, are there papers or books about the architacture and design of CppAD?

**Comment by bradbell** [comment, marked as answer]
Matrix factorization is an advanced AD topic.
Here are a couple of links to papers on this topic:
https://link.springer.com/chapter/10.1007/978-3-540-68942-3_4
https://www.sciencedirect.com/science/article/abs/pii/S1877750311001013

There is an example of atomic Cholesky factoriation at
https://coin-or.github.io/CppAD/doc/atomic_two_eigen_cholesky.cpp.htm
This uses the old atomic_two interface and I have not had time to convert it to use the newer atomic four interface.

It my be better (and much easier) to let eigen do the work
https://coin-or.github.io/CppAD/doc/eigen.htm

Although there is a problem with eigen that the I have had to work around:
https://eigen.tuxfamily.org/bz/show_bug.cgi?id=1480
